### ISSUE TITLE
Human bot aiming issues

### ISSUE THREAD

**Issue by VReaperV** [Unvanquished/Unvanquished#2732]
Human bots currently have extremely bad aim in most situations, resulting in one-sided match-ups, since alien bots (and psaw bots for that matter) virtually never miss.

This layout was built to illustrate the issue:
![unvanquished_2023-06-19_173958_000](https://github.com/Unvanquished/Unvanquished/assets/10687142/131534ff-62f2-4657-89ee-bb2d71871292)
And changed to this later to let aliens evolve:
![unvanquished_2023-06-19_205434_000](https://github.com/Unvanquished/Unvanquished/assets/10687142/270a2783-2122-412f-b552-68563c96037f)

The following cvars were set: `g_bot_attackStruct off; g_bot_infiniteFunds on; g_bot_infiniteMomentum on`. `g_disabledEquipment` and `g_disabledClasses` were set to force bots to use specific weapons/upgrades and alien classes. Bsuit was not allowed. One sk5 bot was spawned in each team for each test.

The human bot's hits and total fired shots were counted for 4 weapons (rifle, lasgun, mdriver and prifle) against 3 classes (dretch, mantis and dragoon) and over multiple attempts. Attempts where one of the bots got stuck were not counted. Below are the tables representing the bot's accuracy for each weapon against each class:

Dretch:
|Weapon|Hit|Fired|Accuracy|
|---|---|---|---|
|Rifle|294|1196|24.58%|
|Lasgun|144|540|26.67%|
|Mdriver|20|110|18.18%|
|Prifle|52|306|16.99%|

Mantis:
|Weapon|Hit|Fired|Accuracy|
|---|---|---|---|
|Rifle|400|1482|26.99%|
|Lasgun|216|798|27.06%|
|Mdriver|35|168|20.83%|
|Prifle|117|393|29.77%|

Dragoon:
|Weapon|Hit|Fired|Accuracy|
|---|---|---|---|
|Rifle|680|1406|48.36%|
|Lasgun|322|564|57.09|
|Mdriver|50|132|37.88%|
|Prifle|299|489|61.15%|

The bots' accuracy is made even worse when going up against players since they strafe. Using adv dragoon in the same area and simply strafing left to right resulted in the average accuracy of a prifle bot of 23.13% (364 hits, 1574 shots fired, and that's the weapon and class that had the best results in the previous testing).

This supports the observations I made while spectating bots on Sweet's server: their accuracy is abysmal, and they often are actively detrimental to their team because they feed evos and momentum to the other team. Oftentimes, I see aliens having a bunch of adv goons when I finish building the base, while h bots, who were told to stay in base, maybe get HS2. The bots' accuracy against smaller aliens is particularly horrible. While they have better accuracy against bigger alien bots, they have the same issues when shooting at the players using bigger classes.

Also, their accuracy with mdriver is especially bad, so getting mass driver over any other weapon is often more detrimental to bots.

**Comment by bmorel** [comment]
it's interesting, because dretch and mantis have the same size, so accuracy against those should be the same, yet for pulse rifle there's a significant difference. Maybe mantis dodge less efficiently.

Against which commit have you done your testing? 0.54 release?

**Comment by VReaperV** [comment]
> it's interesting, because dretch and mantis have the same size, so accuracy against those should be the same, yet for pulse rifle there's a significant difference. Maybe mantis dodge less efficiently.

I think it might be caused by mantis' leap. It leaps while attacking, so the mantis is now a few metres away, moving in a straight line towards the human. So if a bot is firing a pulse rifle it's more likely to hit the mantis. A dretch on the other hand is just gonna run around the human, making it less likely to be hit. Not sure why it only affects the prifle so much, but probably because the other weapons are either slow (lasgun and mass driver), so the human bot can't take as much advantage of this, have spread (rifle), or have less magazine capacity (rifle and mass driver).

> Against which commit have you done your testing? 0.54 release?

0.54 release, yes.

> One thing to know is, I think code was made to make accuracy of aliens and humans different, this might be while aliens almost never miss while humans does that much.
> Also, I don't think SK5 bots should have a too high accuracy. The problem I see in this issue is more that aliens "virtually never miss".

Yeah, that might be the issue. In either case, they just shouldn't be a detriment to their own team or significantly worse than the bots on the other team.

**Comment by sweet235** [comment]
> Also, their accuracy with mdriver is especially bad

I have observed this too. It gets better if you increase the bot skill. However, bot skills of 7 or higher give us rather evil aimbots. I even thought about reducing the likelihood of bots buying mass drivers significantly. 

> 0.54 release

0.54 introduced controversial bot features, in particular preventing them from buying sensible armour randomly. If we only count the accuracy, this should not be too big of a problem. Still, it is neccessary to observe the "bot skilltree skills" as shown by `listbots`. For some reason, these are distributed randomly in 0.54. Since then, the cvars `g_preferredSkillset` and `g_disabledSkillset` were added to allow for working around this problem. But lately, they do not work anymore. I suspect that `g_bot_reactiontime` can influence this measurement too.

Yesterday on IRC, we were talking about if and how we should conduct objective bot tests. The method you present seems to be something we should be doing more often.

**Comment by VReaperV** [comment]
> I have observed this too. It gets better if you increase the bot skill. However, bot skills of 7 or higher give us rather evil aimbots. I even thought about reducing the likelihood of bots buying mass drivers significantly.

Yeah, it feels like they either aim really bad, or almost never miss. I think the difference might be mainly due to predict-aim skill.

> 0.54 introduced controversial bot features, in particular preventing them from buying sensible armour randomly. If we only count the accuracy, this should not be too big of a problem. Still, it is neccessary to observe the "bot skilltree skills" as shown by `listbots`. For some reason, these are distributed randomly in 0.54. Since then, the cvars `g_preferredSkillset` and `g_disabledSkillset` were added to allow for working around this problem. But lately, they do not work anymore. I suspect that `g_bot_reactiontime` can influence this measurement too.

True. The only one that affects aim (for the human bot itself) is predict-aim I think, as I mentioned earlier, maybe movement too, not sure what it does though. The alien bot had the relevant skills, like goon-attack-jump, that could affect the aim of the human bot.

> However, it would seem reasonable to find a method to measure their accuarcy against users too.

I agree. It does introduce more variables though, so the process would probably need to be automated to some degree for that.

> Speaking of, how were those numbers gathered? I'm not aware of any way to gather per-weapon accuracy stats of players (and bots) in unvanquished.

Just counted how much ammo the bot had spent to kill an alien from full health. The bot hit often enough that the regen didn't kick in, and if the human bot died it respawned immediately and bought all the equipment again immediately after respawning. In situations where the alien bot returned to the overmind to heal, or the human bot got stuck, I used admin commands to "reset" them and didn't count the attempt. The only exception was mdriver vs mantis, where the bot would often miss 5-10 times in a row, but I counted the hits there, so it was accounted for.

**Comment by sweet235** [comment]
We removed the random distribution of bot skills in https://github.com/Unvanquished/Unvanquished/pull/2737. By default, `predict-aim` is now given to all bots with skill level 7 and above. Is this appropriate?

**Comment by VReaperV** [comment]
> We removed the random distribution of bot skills in #2737. By default, `predict-aim` is now given to all bots with skill level 7 and above. Is this appropriate?

Yeah, I think this works fine so far. They still suck with particular weapons (mdriver, prifle and lcannon), but that's mostly because they don't know how to use them properly.

**Comment by bmorel** [comment]
another thing which is related on this topic, is that 0.54 increased the default delay that impact their aim.
If I'm not wrong, bots "run" every 100ms (they have 10FPS), and this delay was increased from 100ms for both teams, to 150ms for humans and 250 for aliens. The result is basically that humans have a "maximum simulated lag" of 250ms and aliens of 350ms.

**Comment by sweet235** [comment]
> this delay was increased from 100ms for both teams, to 150ms for humans and 250 for aliens

These are `g_bot_humanAimDelay` and `g_bot_alienAimDelay`. Since 0.54, I have always used 150 for both teams. The actual time is in `sg_bot_util.cpp`:

```
int baseTime = G_Team( self ) == TEAM_ALIENS ? g_bot_alienAimDelay.Get() : g_bot_humanAimDelay.Get();
auto time = ( 10 - self->botMind->skillLevel ) * baseTime * std::max( random(), 0.5f );
```

Changing `baseTime` from 100 to 150 is the difference between skill levels 5 and 7. I think something else was changed in 0.54.

**Comment by bmorel** [comment]
> Changing baseTime from 100 to 150 is the difference between skill levels 5 and 7.

I'm not sure what you mean here? Lowering this number leads to increase of accuracy, for all levels.

**Comment by sweet235** [comment]
(10 - 7) times 150 is approximately (10 - 5) times 100.

**Comment by bmorel** [comment]
This means that current SK5 have roughly the accuracy of previous SK3 bots (since now `(10-5)*150 = 750` and before `(10-7)*100 = 700`)

**Comment by sweet235** [comment]
In 0.54, we also did this: https://github.com/Unvanquished/Unvanquished/commit/33fb1572a739bfc4cd40c70adf92bd0d4aa4f5cb

I think this was the reason to increase the base time.
